compiler/src/iree/compiler/Dialect/Util/IR/UtilAttrs.td

```diff
@@ -104,4 +105,87 @@ def Util_CompositeAttr : AttrDef<Util_Dialect, "Composite", [
   let hasCustomAssemblyFormat = 1;
 }
 
+//===----------------------------------------------------------------------===//
+// Resources
+//===----------------------------------------------------------------------===//
+
+def Util_CasScope : AttrDef<Util_Dialect, "CasScope"> {
+  let mnemonic = "cas-scope";
+  let summary = "A unique scope delimitter for bindings CAS resources";
+  let description = [{
+    CAS resources can be globally (context bound) or locally (sub-tree)
+    bound. Locally scoped resources are eligible for various lifetime
+    policies, liveness analysis, etc, and they can be turned into tombstones
+    if it is deemed valid to do so (releasing the memory they hold while 
+    keeping them in index so an to not produce dangling pointers).
+
+    A scope is demarcated by a #util.cas-scope attribute, which internally
```

Comment:
what does cloning do here? if there was a func.func with a cas scope and a pattern cloned it or created a new one and setDialectAttrs(getDialectAttrs) to preserve the scope attr would that create a new scope or reuse the old one? what would you want to happen? (if we start scoping at the sub-module level we will have this happen, so it'd be good to be clear about that here)

Reply:
Live references to the attribute have pointer equivalence, so clone will reuse the existing scope.